AppDB/appscale/datastore/range_iterator.py

```diff
@@ -0,0 +1,231 @@
+    """
+    # Make sure this filter can be used for a merge join.
+    if pb_filter.op() != Query_Filter.EQUAL:
+      raise BadRequest('Invalid filter for merge join '
+                       '(op must be equal): {}'.format(pb_filter))
+
+    if pb_filter.property_size() != 1:
+      raise BadRequest('Invalid filter for merge join '
+                       '(multiple properties): {}'.format(pb_filter))
+
+    property_ = pb_filter.property(0)
+    if property_.name() == '__key__':
+      raise BadRequest('Invalid property for merge join '
+                       '(must not be __key__): {}'.format(property_))
+
+    return cls(db, project_id, namespace, kind, property_.name(),
+               property_.value())
+
+  @staticmethod
+  def entry_from_result(result):
+    """ Creates an IndexEntry from a Cassandra result.
+
+    Args:
+      result: A dictionary mapping a Cassandra key to a reference value.
+    Returns:
+      An IndexEntry.
+    """
+    entry_key = result.keys()[0]
+    encoded_path = entry_key.rsplit(KEY_DELIMITER)[-1]
+    path = decode_path(encoded_path)
+    entity_ref = result.values()[0]['reference']
+    return IndexEntry(encoded_path, entity_ref, entry_key, path)
+
+  def get_cursor(self):
+    """ Fetches the range's current cursor position.
+
+    Returns:
+      An entity_pb.Path object.
+    """
+    # If the current cursor does not have a path, return an empty one.
+    if self._cursor.key == self.prefix:
+      return Path()
+
+    encoded_path = self._cursor.key.rsplit(KEY_DELIMITER)[-1]
+    return decode_path(encoded_path)
+
+  def set_cursor(self, path, inclusive):
+    """ Changes the range's cursor position.
+
+    Args:
+      path: An entity_pb.Path object.
+      inclusive: A boolean specifying that the next result can include the
+        given path.
+    Raises:
+      BadRequest if unable to set the cursor to the given path.
+    """
+    range_start, range_end = self._range
+    cursor = Cursor(
+      KEY_DELIMITER.join([range_start, str(encode_index_pb(path))]), inclusive)
+
+    if cursor.key < self._cursor.key:
+      raise BadRequest(
+        'Cursor cannot be moved backwards '
+        '({} < {})'.format(repr(cursor.key), repr(self._cursor.key)))
+
+    if cursor.key > range_end:
+      raise BadRequest('Cursor outside range: {}'.format(self._range))
+
+    self._cursor = cursor
+
+  def restrict_to_path(self, path):
+    """ Narrows the range to a specific entity path.
+
+    Args:
+      path: An entity_pb.Path object.
+    """
+    start_key = KEY_DELIMITER.join([self.prefix, str(encode_index_pb(path))])
+    end_key = ''.join([start_key, TERMINATING_STRING])
+    if start_key < self._range[0] or end_key > self._range[-1]:
+      raise BadRequest('Restriction must be within range')
+
+    if self._cursor.key > end_key:
+      raise BadRequest('Cursor already exceeds new range')
+
+    self._range = (start_key, end_key)
+    self._cursor.key = max(start_key, self._cursor.key)
```

Comment:
isn't self._cursor immutable? (namedtuple)

Reply:
Whoops. You're right. It also looks like we weren't testing that code path. https://github.com/AppScale/hawkeye/pull/140 should cover that case.